src/rust/engine/fs/store/src/snapshot_ops.rs

```diff
@@ -0,0 +1,763 @@
+    )
+    .await?
+    .into_iter()
+    .filter_map(|x| match x {
+      Some((dir, _metadata)) => Some(dir),
+      None => unreachable!(),
+    })
+    .collect();
+    Ok(Self::from_directories(top_level_directories, store))
+  }
+
+  // Flatten file and directory nodes.
+  fn from_directories(top_level_directories: Vec<remexec::Directory>, store: Store) -> Self {
+    let (all_file_nodes, all_directory_nodes): (
+      Vec<remexec::FileNode>,
+      Vec<remexec::DirectoryNode>,
+    ) = top_level_directories.into_iter().fold(
+      (vec![], vec![]),
+      |(mut all_files, mut all_directories),
+       remexec::Directory {
+         files, directories, ..
+       }| {
+        all_files.extend(files.to_vec());
+        all_directories.extend(directories.to_vec());
+        (all_files, all_directories)
+      },
+    );
+    Self::from_flattened_nodes(all_file_nodes, all_directory_nodes, store)
+  }
+
+  // Determine any clashing file or directory nodes.
+  fn from_flattened_nodes(
+    all_file_nodes: Vec<remexec::FileNode>,
+    all_directory_nodes: Vec<remexec::DirectoryNode>,
+    store: Store,
+  ) -> Self {
+    let mut files: IndexMap<PathBuf, Vec<remexec::FileNode>> = IndexMap::new();
+    for file_node in all_file_nodes.into_iter() {
+      let entry = files
+        .entry(PathBuf::from(file_node.get_name()))
+        .or_insert_with(Vec::new);
+      entry.push(file_node);
+    }
+
+    let mut non_clashing_files: Vec<remexec::FileNode> = vec![];
+    let mut clashing_files: IndexMap<PathBuf, Vec<remexec::FileNode>> = IndexMap::new();
+    for (prefix, mut cur_clashing_files) in files.into_iter() {
+      if cur_clashing_files.len() == 1 {
+        non_clashing_files.push(cur_clashing_files.pop().unwrap());
+      } else {
+        assert!(cur_clashing_files.len() > 1);
+        clashing_files.insert(prefix, cur_clashing_files);
+      }
+    }
+
+    let mut directories: IndexMap<PathBuf, Vec<remexec::DirectoryNode>> = IndexMap::new();
+    for directory_node in all_directory_nodes.into_iter() {
+      let entry = directories
+        .entry(PathBuf::from(directory_node.get_name()))
+        .or_insert_with(Vec::new);
+      entry.push(directory_node);
+    }
+
+    let mut non_clashing_directories: Vec<remexec::DirectoryNode> = vec![];
+    let mut clashing_directories: IndexMap<PathBuf, Vec<remexec::DirectoryNode>> = IndexMap::new();
+    for (prefix, mut cur_clashing_directories) in directories.into_iter() {
+      if cur_clashing_directories.len() == 1 {
+        non_clashing_directories.push(cur_clashing_directories.pop().unwrap());
+      } else {
+        assert!(cur_clashing_directories.len() > 1);
+        clashing_directories.insert(prefix, cur_clashing_directories);
+      }
+    }
+
+    DigestMergeContext {
+      non_clashing_files,
+      non_clashing_directories,
+      clashing_files,
+      clashing_directories,
+      store,
+    }
+  }
+
+  pub fn resolve_clashing(
```

Comment:
This method name could be improved, because this method is called regardless of whether there are any clashing entries... it essentially implements the merge.

Unless there is a reason for a consumer of this module to compute a `DigestMergeContext` and then call this method, it probably should not be `pub` though (nor `DigestMergeContext`?)

Reply:
`DigestMergeContext` is not `pub`, which is why this method is, so other structs can use it.

Reply:
Oh! I didn't realize `pub` meant `pub` to the whole module! I will fix that! Thanks!

Reply:
I changed the name of this method to `merge_remaining_colliding_files_and_directories_recursive` and added several lines of documentation describing what it does!